From f2455fc4066731c82f8db5e07f5a6dba47612413 Mon Sep 17 00:00:00 2001
From: Song Bing <b06498@freescale.com>
Date: Wed, 20 May 2015 15:15:08 +0800
Subject: [PATCH 04/68] mpegvideoparse: Need detect picture coding type when
 drain
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Need detect picture coding type when drain

Upstream-Status: Pending

https://bugzilla.gnome.org/show_bug.cgi?id=749617

(cherry picked from commit be626f76286c38435f5cadb76113f1aa0bbd2a4a)
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 gst/videoparsers/gstmpegvideoparse.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/gst/videoparsers/gstmpegvideoparse.c b/gst/videoparsers/gstmpegvideoparse.c
index 849af54..e6b4332 100644
--- a/gst/videoparsers/gstmpegvideoparse.c
+++ b/gst/videoparsers/gstmpegvideoparse.c
@@ -749,6 +749,22 @@ need_more:
   if (GST_BASE_PARSE_DRAINING (parse)) {
     GST_LOG_OBJECT (mpvparse, "draining, accepting all data");
     off = size;
+    /* decide picture codding type */
+    if (mpvparse->pic_offset >= 0 && mpvparse->pic_offset < off) {
+      GstMpegVideoPacket header;
+
+      header.data = map.data;
+      header.type = GST_MPEG_VIDEO_PACKET_PICTURE;
+      header.offset = mpvparse->pic_offset;
+      header.size = map.size - mpvparse->pic_offset;
+      if (gst_mpeg_video_packet_parse_picture_header (&header, &mpvparse->pichdr))
+        GST_LOG_OBJECT (mpvparse, "picture_coding_type %d (%s), ending"
+            "frame of size %d", mpvparse->pichdr.pic_type,
+            picture_type_name (mpvparse->pichdr.pic_type), off - 4);
+      else
+        GST_LOG_OBJECT (mpvparse, "Couldn't parse picture at offset %d",
+            mpvparse->pic_offset);
+    }
     ret = TRUE;
   } else {
     GST_LOG_OBJECT (mpvparse, "need more data");
-- 
2.43.0

